modules/packages/LinearAlgebra.chpl

```diff
@@ -0,0 +1,701 @@
+  if Adom.rank != Bdom.rank then compilerError("Unmatched ranks");
+  if Adom.shape != Bdom.shape then halt("Unmatched shapes");
+  var C: [Adom] eltType = A - B;
+  return C;
+}
+
+/*
+    Generic matrix multiplication, ``A`` and ``B`` can be a matrix, vector, or
+    scalar
+*/
+proc dot(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  // vector-vector
+  if Adom.rank == 1 && Bdom.rank == 1 then
+    return inner(A, B);
+  // matrix-(vector|matrix)
+  else
+    return matMult(A, B);
+}
+
+
+pragma "no doc"
+/* Element-wise scalar multiplication */
+proc dot(A: [?Adom] ?eltType, b) where isNumeric(b) {
+  var C: [Adom] eltType = A*b;
+  return C;
+}
+
+
+pragma "no doc"
+/* Element-wise scalar multiplication */
+proc dot(a, B: []) where isNumeric(a) {
+  return dot(B, a);
+}
+
+
+/* Explicit matrix-(matrix|vector) multiplication */
+private proc matMult(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  // matrix-vector
+  if Adom.rank == 2 && Bdom.rank == 1 then
+    return _matvecMult(A, B);
+  // vector-matrix
+  else if Adom.rank == 1 && Bdom.rank == 2 then
+    return _matvecMult(B, A, trans=true);
+  // matrix-matrix
+  else if Adom.rank == 2 && Bdom.rank == 2 then
+    return _matmatMult(A, B);
+  else
+    compilerError("Rank sizes are not 1 or 2");
+}
+
+
+pragma "no doc"
+/* matrix-vector multiplication */
+private proc _matvecMult(A: [?Adom] ?eltType, X: [?Xdom] eltType, trans=false)
+  where isBLASType(eltType)
+{
+  if Adom.rank != 2 || Xdom.rank != 1 then
+    compilerError("Rank sizes are not 2 and 1");
+  if Adom.shape(2) != Xdom.shape(1) then
+    halt("Mismatched shape in matrix-vector multiplication");
+
+  var op = if trans then BLAS.Op.T
+           else BLAS.Op.N;
+
+  var Y: [Xdom] eltType;
+  gemv(A, X, Y, 1:eltType, 0:eltType, opA=op);
+  return Y;
+}
+
+
+pragma "no doc"
+/* matrix-matrix multiplication */
+private proc _matmatMult(A: [?Adom] ?eltType, B: [?Bdom] eltType)
+  where isBLASType(eltType)
+{
+  if Adom.rank != 2 || Bdom.rank != 2 then
+    compilerError("Rank sizes are not 2");
+  if Adom.shape(2) != Bdom.shape(1) then
+    halt("Mismatched shape in matrix-matrix multiplication");
+
+  var C: [Adom.dim(1), Bdom.dim(2)] eltType;
+  gemm(A, B, C, 1:eltType, 0:eltType);
+  return C;
+}
+
+
+/* Inner product of 2 vectors */
+proc inner(A: [?Adom], B: [?Bdom]) {
+  if Adom.rank != 1 || Bdom.rank != 1 then
+    compilerError("Rank sizes are not 1");
+  if Adom.size != Bdom.size then
+    halt("Mismatched size in vector-vector multiplication");
+
+  return + reduce(A[..]*B[..]);
+}
+
+
+/* Outer product of 2 vectors */
+proc outer(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
+  if Adom.rank != 1 || Bdom.rank != 1 then
+    compilerError("Rank sizes are not 1");
+
+  var C: [{Adom.dim(1), Bdom.dim(1)}] eltType;
+  forall (i,j) in C.domain do
+    C[i, j] = A[i]*B[j];
+  return C;
+}
+
+
+pragma "no doc"
+/* Generic matrix-vector multiplication */
+proc _matvecMult(A: [?Adom] ?eltType, X: [?Xdom] eltType)
```

Comment:
I think this needs a trans option to be consistent with the BLAS matrix-vector piece.